### ISSUE TITLE
Ryuk from testcontainers adoption

### ISSUE THREAD

**Issue by lanwen** [orlangure/gnomock#90]
Testcontainers have a really nice feature to cleanup stuff with help of 
https://github.com/testcontainers/moby-ryuk
https://github.com/testcontainers/testcontainers-go/blob/master/reaper.go

Would you consider that in your library? Or maybe the question one step further - is it possible to adopt low-level testcontainers api under the hood? As I really like api of gnomock, however testcontainer maintainers doing a great job keeping away all the docker issues.

**Comment by orlangure** [comment]
Hi @lanwen, sorry for late response.

Thanks for kind words about Gnomock😻

I like the idea behind `moby-ryuk`, I think it would improve Gnomock user experience, I'll just need to figure out how to address the problem you originally had.

`moby-ryuk` <-> `testcontainers-go` integration has some hard coded values, like delay for killing containers is always 10 seconds, and maybe port number is always assumed to be 8080, but I'm not sure, I just took a quick look at the source code you linked here. These hard coded values are problematic. On the other hand, `moby-ryuk` runs in its own containers and has a suicide timer, which is nice and allows to start this new process independently of the parent process (which runs the tests).

I can't use this project as-is, and need to consider either contributing to it, or forking it to implement missing stuff, or building it from scratch (adding it to Gnomock ecosystem) as the project is relatively small.

Another thing to consider would be whether or not should this feature by enabled by default at some point, potentially breaking existing long-running tests. So either we pick a relatively long default timeout (much more than 10 seconds) and hope nobody runs really slow tests, or we enable it via a new option like `WithContainerSuicide` or something, that should be used explicitly.

What are your thoughts?

**Comment by lanwen** [comment]
Testcontainers team is really nice and open for contributions or even the design changes if it's beneficial for users, so I'd suggest to discuss it directly within the repo first. I'd actually try to merge efforts and projects at some point, just need to figure out how to do that in a smooth way, as I feel it's a waste to maintain 2 parallel projects doing almost the same.

Regarding the adoption for now I like an additional option. However I'm more for the disabling option (means suicide enabled by default). The tool itself as far as I know listens for connection and starts a suicide timer only when it's gone. It could be a goroutine that maintains this connection. When it's gone for some reason - cleanup could happen. 10s in this case should be pretty sufficient.

**Comment by orlangure** [comment]
I took another look at this over the weekend, and at this moment I'm not sure it would be beneficial to begin `testcontainers` low level adoption. It's not the right timing, there are more important things for this project right now. I'm keeping it in mind though, and once the time is right, I'll give it another thought.

Regarding the cleanup, first I'd like to understand the problem better. I assume that the problem is:
- containers are not cleaned up when tests don't live until `gnomock.Stop()` call (after `os.Exit` at some point for example)
- tests leave garbage behind (volumes, networks, images)

I ran a couple of tests and found out that containers, when cleaned up properly using `gnomock.Stop()`, only leave docker images behind. When not cleaned up, containers themselves are not removed, and it is a problem.

I built a small tool similar to `moby-ryuk`, but with simpler API, so I have a solution for containers not being cleaned up in the end. Using this tool, containers will be removed properly, so leftover volumes/networks won't be a problem as well.

Regarding "dangling" images, I'm not sure this is a real problem. Personally, I prefer that the images I use in tests would remain on my machine to speed up tests. Otherwise, I'll have to wait for a complete `docker pull` every time.

# 
So, to summarize: will an additional safety net to clean-up old containers be enough to improve user experience?

**Comment by bsideup** [comment]
@orlangure Out of curiosity - would you say that Ryuk's API isn't simple enough? We tried to keep it as trivial as possible (while keeping all potential use cases in mind), and, frankly speaking, I am surprised to see a new project similar to Ryuk just for the sake of simpler (?) API 😅

**Comment by orlangure** [comment]
Hi @bsideup,

Ryuk's API is pretty simple, but there are a couple of things I'd like to improve:
1. If I could remove the need to implement a custom communication protocol (writing into a connection and waiting for `ACK`), I'd go for it even though this is simple enough.
2. I'd like to have more control over when the containers are stopped. For example, I'd like at least a customizable timeout, or better - a way to kill the container right after the tests are completed even if they take more time than the timeout (10 seconds).
3. Gnomock doesn't use labels (yet) to mark its containers, and I think that using container ids would be simpler than session ids.

>  I am surprised to see a new project similar to Ryuk just for the sake of simpler (?) API 

Yes, that's why I didn't plan to create a whole new project for it. I wrote something very simple, and it was quicker and easier than figuring out how to integrate these changes into Ryuk, assuming you would agree to them (which is not given).

I have an example in mind, and I'll share it soon in a draft pull request in this repo. I implemented only a subset of Ryuk's features, but I believe that would be enough. It would be great if you could share your thoughts about it.

**Comment by orlangure** [comment]
@lanwen, I merged container cleanup mechanism, for now without using adding any new dependencies. Again, this might change at some point.
